src/status.ts

```diff
@@ -1,259 +1,454 @@
-    this._debugButton.tooltip = localize('launch.debugger.tooltip', 'Launch the debugger for the selected target');
-    this._debugButton.command = 'cmake.debugTarget';
-    this._launchTargetNameButton.command = 'cmake.selectLaunchTarget';
-    this._launchTargetNameButton.tooltip = localize('select.target.tooltip', 'Select the target to launch');
-    this._reloadBuildButton();
-    this.reloadVisibility();
-  }
-
-  reloadVisibility() {
-    setVisible(this._activeFolderButton, Boolean(this._visible && vscode.workspace.workspaceFolders && vscode.workspace.workspaceFolders.length > 1 && !!this._activeFolderButton.text));
-    const autovis_items = [
-      this._cmakeToolsStatusItem,
-      this._buildButton,
-      this._kitSelectionButton,
-      this._targetButton,
-      this._debugButton,
-      this._launchTargetNameButton,
-    ];
-    for (const item of autovis_items) {
-      setVisible(item, this._visible && !!item.text);
-    }
-    setVisible(this._debugButton,
-               this._visible && !this._hideDebugButton && vscode.extensions.getExtension('ms-vscode.cpptools') !== undefined
-                   && !!this._debugButton.text);
+  get text(): string { return this._text; }
+  set text(v: string) {
+    this._text = v;
+    this.update();
   }
 
-  /**
-   * Whether the status bar items are visible
-   */
-  setVisible(v: boolean) {
-    this._visible = v;
-    this.reloadVisibility();
+  get tooltip(): string|null { return this._tooltip; }
+  set tooltip(v: string|null) {
+    this._tooltip = v;
+    this.update();
   }
-  private _visible: boolean = true;
 
-  private _reloadStatusButton() {
-    this._cmakeToolsStatusItem.text = `CMake: ${this._buildTypeLabel}: ${this._statusMessage}`;
-    this.reloadVisibility();
+  protected set icon(v: string|null) { this._icon = v ? `$(${v})` : null; }
+
+  protected set command(v: string|null) { this.button.command = v || undefined; }
+
+  dispose(): void { this.button.dispose(); }
+  update(): void {
+    const visible = this._isVisible();
+    if (!visible || this._forceHidden) {
+      this.button.hide();
+      return;
+    }
+    const text = this._getText(true);
+    if (text === '') {
+      this.button.hide();
+      return;
+    }
+    this.button.text = text;
+    this.button.tooltip = this._getTooltip() || undefined;
+    this.button.show();
   }
 
-  private _reloadDebugButton() {
-    if (!this._launchTargetNameButton.text) {
-      this._debugButton.text = '$(bug)';
-      this._launchTargetNameButton.hide();
-    } else {
-      this._debugButton.text = `$(bug) ${localize('debug', 'Debug')}`;
-      if (this._visible) {
-        this._launchTargetNameButton.show();
-      }
+  protected getTextNormal(): string { return this._text; }
+  protected getTextShort(): string { return this.getTextNormal(); }
+  protected getTextIcon(): string { return ''; }
+
+  protected getTooltipNormal(): string|null { return this._tooltip; }
```

Comment:
Tooltip issues:

**`default` mode**
Button|comment
---|---
**Status**| Remove the status from the tooltip. It should just say "Click to select the current build variant"

**`compact` mode**
Button|comment
---|---
**Workspace**| I don't think we need to show the folder name in the tooltip because we don't truncate the text.
**Status**| I think the status should come before the command text. Can you switch the order?
**Kit**| I think the active kit should come before the command text. Can you switch the order?
**Build**| The text should start with "CMake:"
**Build label**| I don't think we need to show the selected build in the tooltip because we don't truncate the text.
**Launch label**| The text should start with "CMake:"
**CTest**| I don't think we need to show the button text in the tooltip. (make it a one-liner)

**`icon` mode**
Button|comment
---|---
**Workspace**| I'm thinking this should look like: "CMake: Active Folder [${folder name}]\nClick to change"
**Status**| I'm thinking this should look like: "CMake: ${status text}\nClick to select the current build variant" (In the future, maybe status and variant will be split out so we can put the variant in square brackets, but I think that's out of scope for this PR)
**Kit**| I'm thinking this should look like: "CMake: Active Kit [${active kit}]\nClick to change the active kit"
**CTest**| I don't think we need to show the button text in the tooltip. (make it a one-liner)

Reply:
ctest in icon mode should have the annotation of how many test did  pass.
=> second line is only visible if there are test results

i changed the labels (and tooltips) such that selected values (kit, folder, variant, targets) are in brackets and text is without bracket.

i didn't add Active XYZ to icon mode.
I don't think that it adds information because the second line of the tooltip already states what is meant.

I changed the tooltip of the workspace button to:
`Click to select the active workspace` and `Click to select the active workspace (auto)` to match the other tooltips.
i'm not sure if auto is enough or if we should use autoselect.
auto is visible if Cmake: Auto Select Active Folder is selected.

Reply:
I think putting (auto) in the tooltip is confusing.  I would rather that be removed.